feat(apigateway): step functions integration (#16827)
aws/aws-cdk: packages/@aws-cdk/aws-apigateway/README.md

The following code defines a REST API that routes all requests to the specified AWS StepFunctions state machine:

```ts
const stateMachineDefinition = new stepfunctions.Pass(this, 'PassState');

const stateMachine: stepfunctions.IStateMachine = new stepfunctions.StateMachine(this, 'StateMachine', {
    definition: stateMachineDefinition,
    stateMachineType: stepfunctions.StateMachineType.EXPRESS,
});
    
new apigateway.StepFunctionsRestApi(this, 'StepFunctionsRestApi', {
  deploy: true,
  stateMachine: stateMachine,
});
```

Here are a few examples:

Example 1: POST with default configuration

```bash
curl -X POST -d '{ "customerId": 1 }' https://example.com/
```

AWS Step Functions will receive the request body inside a `body` key as it is possible to have other keys within the request (ex. requestContext, header, path, and/or querystring):

```json
{
  "body": {
    "customerId": 1 
  },
  "path": "/",
  "querystring": {}
}
```

Example 2: GET with path `/users/{userId}`.

```bash
curl -X GET https://example.com/users/{userId}
```

AWS Step Functions will receive the following execution input:

```json
{
  "body": {},
  "path": "/users/{userId}",
  "querystring": {}
}

> ```

Reviewer: updated. I changed to path to be a json object so users can access it from jsonpath within the state machine definition.

  deploy: true,
  stateMachine: stateMachine,
});
```

When the REST API endpoint configuration above is invoked using POST, as follows -

```bash
curl -X POST -d '{ "customerId": 1 }' https://example.com/
```

AWS Step Functions will receive the request body in its input as follows:

```json
{
  "body": {
    "customerId": 1 
  },
  "path": "/",
  "querystring": {}
}
```

When the endpoint is invoked at path '/users/5' using the HTTP GET method as below:

```bash
curl -X GET https://example.com/users/5?foo=bar
```

AWS Step Functions will receive the following execution input:

```json
{
  "body": {},
  "path": {
     "users": "5"
  },
  "querystring": {
    "foo": "bar"
  }
}
```

It is possible to combine included/excluded fields. For example, `headers` and the `requestContext` can be included while `path` and `querystring` are excluded through the following configuration:

```ts
const stateMachineDefinition = new stepfunctions.Pass(this, 'PassState');

const stateMachine: stepfunctions.IStateMachine = new stepfunctions.StateMachine(this, 'StateMachine', {
    definition: stateMachineDefinition,
    stateMachineType: stepfunctions.StateMachineType.EXPRESS,
});

>     

Reviewer 2: I've pushed a commit that fixes this up.

>     

Reviewer: Thank you!

new apigateway.StepFunctionsRestApi(this, 'StepFunctionsRestApi', {
  stateMachine: stateMachine,
  headers: true,
  path: false,
  querystring: false,
  requestContext: {
      caller: true,
      user: true,
  },
});
```

> When the endpoint is invoked at path '/' using the HTTP GET method as below:

Reviewer 2: ```suggestion
In such a case, when the endpoint is invoked as below:
```

item.addMethod('DELETE', new apigateway.HttpIntegration('http://amazon.com'));
```

## AWS StepFunctions backed APIs

You can use Amazon API Gateway with AWS Step Functions as the backend integration, specifically Synchronous Express Workflows.

The `StepFunctionsRestApi` construct makes this easy and also sets up input, output and error mapping. The `StepFunctionsRestApi` construct sets up the API Gateway REST API with an `ANY` HTTP method and sets up the api role with the required permission to invoke `StartSyncExecution` action on the AWS StepFunctions state machine. This will enable you to invoke any one of the API Gateway HTTP methods and get a response from the backend AWS StepFunctions state machine. Invoking any of the HTTP Method, using the example below, will return a "Hello!" message as the Response Body.

> The following code defines a REST API that routes all requests to the specified AWS StepFunctions state machine:

Reviewer 2: My comment is not so much focused on stages here, but resources and methods - https://docs.aws.amazon.com/apigateway/latest/developerguide/how-to-method-settings.html

item.addMethod('DELETE', new apigateway.HttpIntegration('http://amazon.com'));
```

## AWS StepFunctions backed APIs

You can use Amazon API Gateway with AWS Step Functions as the backend integration, specifically Synchronous Express Workflows.

The `StepFunctionsRestApi` construct makes this easy and also sets up input, output and error mapping. The `StepFunctionsRestApi` construct sets up the API Gateway REST API with an `ANY` HTTP method and sets up the api role with the required permission to invoke `StartSyncExecution` action on the AWS StepFunctions state machine. This will enable you to invoke any one of the API Gateway HTTP methods and get a response from the backend AWS StepFunctions state machine.

The following code defines a REST API that routes all requests to the specified AWS StepFunctions state machine:

```ts
const stateMachine = new stepFunctions.StateMachine(this, 'StateMachine', ...);
new apigateway.StepFunctionsRestApi(this, 'StepFunctionsRestApi', { 
  stateMachine: stateMachine,
});

> ```

Reviewer 2: I'm still finding it hard to understand.

The paragraph is describing the setup instead it should explain the feature.

Can you improve this example with a simple but actual state machine and explain what invoking the different APIGateway routes would do?

Reviewer 2: We have a pretty nice (unfortunately, undocumented) feature that allows you to insert snippets from an integration test (`integ.*.lit.ts`) directly into the docs.

See https://github.com/aws/aws-cdk/blob/7966f8d48c4bff26beb22856d289f9d0c7e7081d/packages/%40aws-cdk/aws-logs/test/example.retention.lit.ts#L7-L22
and see the README - https://github.com/aws/aws-cdk/blob/7966f8d48c4bff26beb22856d289f9d0c7e7081d/packages/%40aws-cdk/aws-logs/README.md#log-groupsstreams

This finally gets rendered into https://docs.aws.amazon.com/cdk/api/latest/docs/aws-logs-readme.html#log-groupsstreams

It might be worth using that feature here.

You can add requestContext (similar to input requestContext from lambda input) to the input. The 'requestContext' parameter includes account ID, user identity, etc. that can be used by customers that want to know the identity of authorized users on the state machine side. The following code defines a REST API like above but also adds 'requestContext' to the input of the State Machine:

```ts
const stateMachine = new stepFunctions.StateMachine(this, 'StateMachine', ...);
new apigateway.StepFunctionsRestApi(this, 'StepFunctionsRestApi', { 
  stateMachine: stateMachine,
  includeRequestContext: true,
});

> ```

Reviewer 2: Move this explanation to the docs of `requestContext` and drop all of this here.

item.addMethod('DELETE', new apigateway.HttpIntegration('http://amazon.com'));
```

## AWS StepFunctions backed APIs

You can use Amazon API Gateway with AWS Step Functions as the backend integration, specifically Synchronous Express Workflows.

> The `StepFunctionsRestApi` only supports integration with Synchronous Express state machine. The `StepFunctionsRestApi` construct makes this easy and also sets up input, output and error mapping. The `StepFunctionsRestApi` construct sets up the API Gateway REST API with an `ANY` HTTP method and sets up the api role with the required permission to invoke `StartSyncExecution` action on the AWS StepFunctions state machine. It sets up up a `prod` stage by default. This will enable you to invoke any of the API Gateway HTTP methods for that `prod` and get a response from the backend AWS Step Functions execution. Invoking either GET or POST in the example below will send the request to the state machine as a new execution. On success, an HTTP code '200' is returned with ONLY the execution output as the Response Body. If the state machine execution fails, an HTTP '500' error response is returned with the error and cause of the execution as the Response Body. If the request is invalid (ex. bad execution input) an HTTP '400' error is returned.

Reviewer: Updated to:

```
The construct sets up an API endpoint and maps the `ANY` HTTP method and any calls to the API endpoint starts an express workflow execution for the underlying state machine. 
```

As part of the API Gateway integration with Step Functions, it is possible to opt-in to include requestContext, headers, path, and querystring to the execution input. By default, these paramaters are not included in order to reduce payload size sent to Step Functions. 

> More details about AWS Step Functions payload limit can be found at https://docs.aws.amazon.com/step-functions/latest/dg/limits-overview.html#service-limits-task-executions.

Reviewer: updated

The following code defines a REST API that routes all requests to the specified AWS StepFunctions state machine:

```ts
const machineDefinition = new sfn.Pass(this, 'PassState', {

> })

Reviewer: updated

const stateMachine: sfn.IStateMachine = new sfn.StateMachine(this, 'StateMachine', {
    definition: machineDefinition,
    stateMachineType: sfn.StateMachineType.EXPRESS,
});
    
new apigateway.StepFunctionsRestApi(this, 'StepFunctionsRestApi', { 
  stateMachine: stateMachine,
});
```

Here are a few examples:

Example 1: POST with default configuration

> ```json
> POST /
> {
>      "customerId": 1 
> }
> ```

Reviewer 2: ```suggestion
curl -X POST -d '{ "customerId": 1 }' https://example.com/
```

> ```

Reviewer: definitely, I will update the examples.

> Step Functions will wrap the request body inside a `body` key as it is possible to have other keys within the request (ex. requestContext, header, path, and/or querystring):

Reviewer: It is accurate but I will remove to avoid confusion. The point here was to clarify that a request body will not be the whole execution input.

an API request with this body:

```json
{
    "customerId": 1
}
```

will be translated to:

```json
{
   "body": {
        "customerId": 1
   } 
}
```

as API Gateway can forward more fields (headers, path, requestContext) and they are not part of the request body.

> Step Functions will wrap the request body inside a `body` key as it is possible to have other keys within the request (ex. requestContext, header, path, and/or querystring):

Reviewer: hmm.. it depends on the execution output. the only mapping for the response is to return only the execution output (execution ARN, state machine ARN, account id, billing data, etc) are removed. Users don't usually expose those data when they offer to start an execution from an endpoint. 

I will clarify that
